### PR TITLE
[MM-36240] Add commands to enable/disable apps

### PR DESCRIPTION
**Pull request by hanzei**
#### Summary
This PR adds a `/apps enable` and `/apps disable` command.

As discussed in https://community-daily.mattermost.com/core/pl/onpuah5sx3gipxhgqw85o7arza the life cycle behaviors is as following:

Disable:
- Disable Bot account. This also renders the bot access token unless until the bot is enabled again.
- Mark the app as disabled, therefore skipping in when bindings are fetched.
- The OAuth app _should_ also be marked as disabled, but the server doesn't support that. Filled https://mattermost.atlassian.net/browse/MM-36337 to work on that in the future.
- _Not_ delete the KV data

Uninstall:
- Disable Bot account and revoke the bot access token.
- Delete the OAuth app
- Delete the KV data of the apps

The behavior of the life cycle methods is also now unified. All four (enable, disable, install, uninstall) only give the app a call if there is a call defined in the manifest. The app can overwrite the default message send to the user as an ephemeral post via `Markdown`.

Please note that the `DeleteAll` method is far from optimized. I didn't found a way to make perform better without server changes. 

There is a also a `hello-lifecycle` app that I used for testing.

This is now an app is marked in the Marketplace:
Screenshot from 2021-06-10 12-47-32

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-36240
https://mattermost.atlassian.net/browse/MM-36335

Fixes https://github.com/mattermost/mattermost-plugin-apps/issues/208

### CODE REVIEW

File: apps/mmclient/mattermost_client.go [hunk 1]
```diff
@@ -160,16 +160,16 @@ func (c *Client) CreatePost(post *model.Post) (*model.Post, error) {
 	return createdPost, nil
 }
 
-func (c *Client) DM(userID string, format string, args ...interface{}) {
+func (c *Client) DM(userID string, format string, args ...interface{}) (*model.Post, error) {
```
Comment (hanzei) [inline apps/mmclient/mattermost_client.go:163]: It's useful for debugging return an error.

File: go.mod [hunk 2]
```diff
@@ -3,7 +3,9 @@ module github.com/mattermost/mattermost-plugin-apps
 go 1.16
 
 require (
+	github.com/aws/aws-lambda-go v1.19.1
```
Comment (hanzei) [inline go.mod:6]: My editor didn't liked the nested go modules files and hence I move the two dependencies back into the root one. IMO it makes thing simpler.

File: server/proxy/install.go [hunk 3]
```diff
@@ -189,34 +187,21 @@ func (p *Proxy) ensureBot(app *apps.App, actingUserID string, client *model.Clie
...
+	// Create an access token on a fresh app install
+	if app.RequestedPermissions.Contains(apps.PermissionActAsBot) &&
+		app.BotAccessTokenID == "" {
+		token, response := client.CreateUserAccessToken(fullBot.UserId, "Mattermost App Token")
+		if response.Error != nil {
+			return errors.Wrap(response.Error, "failed to create bot user's access token")
+		}
+		if response.StatusCode != http.StatusOK {
+			return fmt.Errorf("failed to create bot user's access token, status code = %v", response.StatusCode)
 		}
```
Comment (hanzei) [inline server/proxy/install.go:199]: This part fixes https://mattermost.atlassian.net/browse/MM-36335.

`GetUserAccessToken` doesn't return `token.Token` for security reasons and hence `app.BotAccessToken` will be empty. Given that there is no need to fetch the token again (it's in the store)  the code only creates one on a fresh install.

File: server/proxy/install.go [hunk 4]
```diff
@@ -189,34 +187,21 @@ func (p *Proxy) ensureBot(app *apps.App, actingUserID string, client *model.Clie
...
+
 		app.BotAccessToken = token.Token
 		app.BotAccessTokenID = token.Id
 	}
 
-	_ = p.mm.Post.DM(fullBot.UserId, actingUserID, &model.Post{
-		Message: fmt.Sprintf("Using bot account @%s (`%s`).",
-			fullBot.Username, fullBot.UserId),
-	})
-
```
Comment (hanzei) [inline server/proxy/install.go:219]: This was broken anyway.

File: server/proxy/uninstall.go [hunk 5]
```diff
@@ -4,83 +4,98 @@
...
-			return errors.Errorf("failed to revoke bot access token for %s, returned with status code %d", app.AppID, response.StatusCode)
+			return "", errors.Errorf("failed to revoke bot access token for %s, returned with status code %d", app.AppID, response.StatusCode)
 		}
 	}
 
-	_, err = p.mm.Bot.UpdateActive(app.BotUserID, false)
-	if err != nil {
-		return errors.Wrapf(err, "failed to disable bot account for %s", app.AppID)
+	// disable the bot account
+	_, response := asAdmin.DisableBot(app.BotUserID)
```
Comment (hanzei) [inline server/proxy/uninstall.go:74]: `DisableBot` is the REST API equivalent of`p.mm.Bot.UpdateActive`.

**Review by levb — COMMENTED** [review]
Nice, LGTM

**Review by larkox — APPROVED** [review]

**Comment by hanzei** [comment]
@DHaussermann There are no clear QA steps here. My advice would be to play around with `hello-lifecycle`  and the other apps.

**Comment by hanzei** [comment]
@levb Would you please merge this PR after QA is complete?

**Comment by levb** [comment]
@DHaussermann What is the status of testing this PR?

**Review by DHaussermann — APPROVED** [review]
Tested and passed 
- No Regressions found in install / uninstall process
- Apps shows correctly as disabled in apps list table when this is the case
- Disabling an app disables its respective bot
- KV store data is preserved as expected and available when an app is re-enabled 
- Feedback is clear when running various invalid forms of the enable / disable commands

LGTM!

**Comment by DHaussermann** [comment]
cc @levb ☝️ Please merge when you have a chance as Ben is away.

**Pull request merged**